### PR TITLE
Prevent upgrades between 'nightly' and 'release' builds of Dotnet CLI.

### PR DESCRIPTION
**Pull request by Sridhar-MS**
Set a new registry 'BuildType' when installing. Check for this reg key
when upgrading to a newer version. Show error message and exit
if the previous installation does not have the same 'BuildType'.

/cc @piotrpMSFT 

image

### CODE REVIEW

File: packaging/windows/variables.wxi [hunk 1]
```diff
@@ -7,13 +13,14 @@
...
   <?define Manufacturer     =   "Microsoft Corporation" ?>
-  <?define ProductName      =   "Microsoft Dotnet CLI for Windows ($(var.DisplayVersion))" ?>
+  <?define ProductName      =   "Microsoft Dotnet CLI for Windows ($(var.DisplayVersion) $(var.ReleaseSuffix))" ?>
   <?define ProductLanguage  =   "1033" ?>
   <?define ProductVersion   =   "$(var.Dotnet_ProductVersion)" ?>
   <?define ProductFamily    =   "dotnet" ?>
   <?define ProductEdition   =   "001dotnet" ?>
   <?define LCID  = "$(var.ProductLanguage)"?>
   <?define DowngradeErrorMessage  = "A newer version is already installed; please uninstall it and re-run setup."?>
+  <?define IncompatibleBuildTypeError = "You have installed '[INSTALLEDCLIBUILDTYPE]' build of Dotnet CLI. It is cannot be upgraded to a '[EXPECTEDCLIBUILDTYPE]' build of Dotnet CLI. Please uninstall it and re-run setup."?>
```
Comment (shahid-pk) [inline packaging/windows/variables.wxi]: Nit : "It is cannot be upgraded" should be "It cannot be upgrade" i think.

**Comment by davidfowl** [comment]
Why is this a good idea? Do we know if other platforms prevent this?

**Comment by Sridhar-MS** [comment]
@dotnet-bot test this please.

File: packaging/windows/variables.wxi (hunk 1 above)
Comment (Sridhar-MS) [inline packaging/windows/variables.wxi]: Its a typo. I will fix it.

**Comment by Sridhar-MS** [comment]
@davidfowl This is mainly because how we want to do branching for release. The master and the release branch are individually guaranteed to have increasing version numbers. But we do not make that promise across the branches. So a major upgrade from `x.y` to `x.y1` may not provide backward compatibility depending on where `x.y` or `x.y1` came from (ie. release or nightly).

**Comment by Sridhar-MS** [comment]
@dotnet-bot  test CentOS7.1 Release Build please

**Comment by Sridhar-MS** [comment]
@dotnet-bot test CentOS7.1 Debug Build please

**Comment by Sridhar-MS** [comment]
@dotnet-bot test Ubuntu Debug Build please

**Comment by Sridhar-MS** [comment]
@dotnet-bot test Ubuntu Release Build please

**Comment by piotrpMSFT** [comment]
@dotnet-bot test Ubuntu Debug Build please.

File: packaging/windows/variables.wxi [hunk 2]
```diff
@@ -1,5 +1,11 @@
 <?xml version="1.0"?>
 <Include xmlns="http://schemas.microsoft.com/wix/2006/wi">
+  <?if $(var.ReleaseSuffix) ~= "beta" OR $(var.ReleaseSuffix) ~= "rc1" OR $(var.ReleaseSuffix) ~= "rtm" OR $(var.ReleaseSuffix) ~= ""?>
+  <?define BuildType="Release"?>
```
Comment (piotrpMSFT) [inline packaging/windows/variables.wxi]: ok for now, but we'll need to keep updating this. Let's make this a boolean field that denotes we're in the release/prod train. Simple reg key?

Comment (Sridhar-MS) [inline packaging/windows/variables.wxi]: I can move this logic to the generatemsi.ps1 script and pass the `BuildType` as a variable to the wix files.

Comment (Sridhar-MS) [inline packaging/windows/variables.wxi]: Created issue - #878

**Pull request merged**